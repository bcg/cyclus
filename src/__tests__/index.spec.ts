import { Lifecycle, SystemMap, using } from "..";

describe("cyclus", () => {
  function timeout(ms: number): Promise<{}> {
    return new Promise((resolve) => {
      setTimeout(resolve, ms);
    });
  }

  class Database extends Lifecycle {
    public dbConnection: string;

    public start() {
      order.push("Start Database");
      this.dbConnection = "OPENED";
    }

    public stop() {
      order.push("Stop Database");
      this.dbConnection = "CLOSED";
    }
  }

  class Scheduler extends Lifecycle {
    public tick: number;

    public start() {
      order.push("Start Scheduler");
      this.tick = 10;
    }

    public stop() {
      order.push("Stop Scheduler");
    }
  }

  class ExampleComponent extends Lifecycle {
    private database: Database;
    private scheduler: Scheduler;

    public start() {
      order.push("Start ExampleComponent");
    }

    public stop() {
      order.push("Stop ExampleComponent");
    }
  }

  class NewDatabase extends Lifecycle {
    public dbConnection: string;

    public start() {
      order.push("Start NewDatabase");
      this.dbConnection = "OPENED";
    }

    public stop() {
      order.push("Stop NewDatabase");
      this.dbConnection = "CLOSED";
    }
  }

  const createDatabase = (): Lifecycle => new Database();
  const createNewDatabase = (): Lifecycle => new NewDatabase();
  const createScheduler = (): Lifecycle => new Scheduler();
  const createExampleComponent = (): Lifecycle => new ExampleComponent();

  let order: string[];
  let system: SystemMap;

  beforeEach(() => {
    order = [];
  });

  describe("lifecycle", () => {
    beforeEach(async () => {
      system = new SystemMap({
        database: createDatabase(),
        scheduler: createScheduler(),
        exampleComponent: using(createExampleComponent(), [
          "database",
          "scheduler"
        ])
      });
      await system.start();
    });

    it("should start the system in order", () => {
      expect(order).toMatchSnapshot();
    });

    it("should stop the system in reversed order", async () => {
      await system.stop();
      expect(order).toMatchSnapshot();
    });

    it("starting should be idempotent", async () => {
      await system.start();
      expect(order).toMatchSnapshot();
    });

    it("stopping should be idempotent", async () => {
      await system.stop();
      await system.stop();
      expect(order).toMatchSnapshot();
    });
  });

  describe("injecting dependencies correctly", () => {
    it("should work with map", async () => {
      system = new SystemMap({
        db: createDatabase(),
        sched: createScheduler(),
        exampleComponent: using(createExampleComponent(), {
          database: "db",
          scheduler: "sched"
        })
      });

      await system.start();

      expect(system).toMatchSnapshot();
    });
  });

  describe("replacing dependencies on the fly", () => {
    it("should replace but not start or stop any component", async () => {
      system = new SystemMap({
        database: createDatabase(),
        scheduler: createScheduler(),
        exampleComponent: using(createExampleComponent(), [
          "database",
          "scheduler"
        ])
      });

      expect(system).toMatchSnapshot();
      await system.replace({ database: createNewDatabase() });
      expect(system).toMatchSnapshot();

      expect(order).toMatchSnapshot();
    });

    it("should work for test use case", async () => {
      system = new SystemMap({
        database: createDatabase(),
        scheduler: createScheduler(),
        exampleComponent: using(createExampleComponent(), [
          "database",
          "scheduler"
        ])
      });

      await system.replace({ database: createNewDatabase() });
      await system.start();

      expect(order).toMatchSnapshot();
    });

    it("should restart correctly", async () => {
      system = new SystemMap({
        database: createDatabase(),
        scheduler: createScheduler(),
        exampleComponent: using(createExampleComponent(), [
          "database",
          "scheduler"
        ])
      });

      await system.start();

      expect(system).toMatchSnapshot();
      await system.replace({ database: createNewDatabase() }, { shouldRestart: true });
      expect(system).toMatchSnapshot();

      expect(order).toMatchSnapshot();
    });
  });

  describe("async", () => {
    class Component1 extends Lifecycle {
      public async start() {
        await timeout(100);
        order.push("Start component 1 after 1000");
      }

      public async stop() {
        await timeout(100);
        order.push("Stop component 1 after 1000");
      }
    }

    class Component2 extends Lifecycle {
      public async start() {
        await timeout(200);
        order.push("Start component 2 after 2000");
      }

      public async stop() {
        await timeout(200);
        order.push("Stop component 2 after 2000");
      }
    }

    class Component3 extends Lifecycle {
      private component1: Component1;
      private component2: Component2;

      public start() {
        order.push("Start component 3");
      }

      public stop() {
        order.push("Stop component 3");
      }
    }

    const createComponent1 = (): Lifecycle => new Component1();
    const createComponent2 = (): Lifecycle => new Component2();
    const createComponent3 = (): Lifecycle => new Component3();

    it("should wait for start or stop method to resolve", async () => {
      system = new SystemMap({
        component1: createComponent1(),
        component2: createComponent2(),
        component3: using(createComponent3(), ["component1", "component2"])
      });

      await system.start();

      expect(system).toMatchSnapshot();
      expect(order).toMatchSnapshot();
    });
  });

  describe("simple component", () => {
    class Component extends Lifecycle {
      private config: object;

      public start() {
        order.push("Start component 3");
      }

      public stop() {
        order.push("Stop component 3");
      }
    }
    const createComponent = (): Lifecycle => new Component();

    beforeEach(async () => {
      system = new SystemMap({
        config: {
          a: 1,
          b: 2
        },
        component: using(createComponent(), ["config"])
      });
      await system.start();
    });

    it("should work correctly", () => {
      expect(system).toMatchSnapshot();
    });

    it("should able to replace the config on the fly", async () => {
      await system.replace({
        config: { c: 3, d: 4 }
      });
      expect(system).toMatchSnapshot();
    });
  });
});
