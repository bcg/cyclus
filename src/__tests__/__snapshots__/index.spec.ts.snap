// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cyclus async 'should wait for start or stop method to resolve 1`] = `
SystemMap {
  "__metadata": Object {
    "__cache": Object {
      "@cyclus/SystemMap/BUILT_ORDER": Array [
        "component2",
        "component1",
        "component3",
      ],
    },
  },
  "map": Object {
    "component1": Component1 {
      "__metadata": Object {
        "dependencies": Object {},
        "isInitialised": true,
      },
    },
    "component2": Component2 {
      "__metadata": Object {
        "dependencies": Object {},
        "isInitialised": true,
      },
    },
    "component3": Component3 {
      "__metadata": Object {
        "dependencies": Object {
          "component1": "component1",
          "component2": "component2",
        },
        "isInitialised": true,
      },
      "component1": Component1 {
        "__metadata": Object {
          "dependencies": Object {},
          "isInitialised": true,
        },
      },
      "component2": Component2 {
        "__metadata": Object {
          "dependencies": Object {},
          "isInitialised": true,
        },
      },
    },
  },
}
`;

exports[`cyclus async 'should wait for start or stop method to resolve 2`] = `
Array [
  "Start component 2 after 2000",
  "Start component 1 after 1000",
  "Start component 3",
]
`;

exports[`cyclus injecting dependencies correctly 'should work with map 1`] = `
SystemMap {
  "__metadata": Object {
    "__cache": Object {
      "@cyclus/SystemMap/BUILT_ORDER": Array [
        "sched",
        "db",
        "exampleComponent",
      ],
    },
  },
  "map": Object {
    "db": Database {
      "__metadata": Object {
        "dependencies": Object {},
        "isInitialised": true,
      },
      "dbConnection": "OPENED",
    },
    "exampleComponent": ExampleComponent {
      "__metadata": Object {
        "dependencies": Object {
          "database": "db",
          "scheduler": "sched",
        },
        "isInitialised": true,
      },
      "database": Database {
        "__metadata": Object {
          "dependencies": Object {},
          "isInitialised": true,
        },
        "dbConnection": "OPENED",
      },
      "scheduler": Scheduler {
        "__metadata": Object {
          "dependencies": Object {},
          "isInitialised": true,
        },
        "tick": 10,
      },
    },
    "sched": Scheduler {
      "__metadata": Object {
        "dependencies": Object {},
        "isInitialised": true,
      },
      "tick": 10,
    },
  },
}
`;

exports[`cyclus lifecycle should start the system in order 1`] = `
Array [
  "Start Scheduler",
  "Start Database",
  "Start ExampleComponent",
]
`;

exports[`cyclus lifecycle should stop the system in reversed order 1`] = `
Array [
  "Start Scheduler",
  "Start Database",
  "Start ExampleComponent",
  "Stop ExampleComponent",
  "Stop Database",
  "Stop Scheduler",
]
`;

exports[`cyclus lifecycle starting should be idempotent 1`] = `
Array [
  "Start Scheduler",
  "Start Database",
  "Start ExampleComponent",
]
`;

exports[`cyclus lifecycle stopping should be idempotent 1`] = `
Array [
  "Start Scheduler",
  "Start Database",
  "Start ExampleComponent",
  "Stop ExampleComponent",
  "Stop Database",
  "Stop Scheduler",
]
`;

exports[`cyclus replacing dependencies on the fly 'should work correctly 1`] = `
SystemMap {
  "__metadata": Object {
    "__cache": Object {
      "@cyclus/SystemMap/BUILT_ORDER": Array [
        "scheduler",
        "database",
        "exampleComponent",
      ],
    },
  },
  "map": Object {
    "database": Database {
      "__metadata": Object {
        "dependencies": Object {},
        "isInitialised": true,
      },
      "dbConnection": "OPENED",
    },
    "exampleComponent": ExampleComponent {
      "__metadata": Object {
        "dependencies": Object {
          "database": "database",
          "scheduler": "scheduler",
        },
        "isInitialised": true,
      },
      "database": Database {
        "__metadata": Object {
          "dependencies": Object {},
          "isInitialised": true,
        },
        "dbConnection": "OPENED",
      },
      "scheduler": Scheduler {
        "__metadata": Object {
          "dependencies": Object {},
          "isInitialised": true,
        },
        "tick": 10,
      },
    },
    "scheduler": Scheduler {
      "__metadata": Object {
        "dependencies": Object {},
        "isInitialised": true,
      },
      "tick": 10,
    },
  },
}
`;

exports[`cyclus replacing dependencies on the fly 'should work correctly 2`] = `
SystemMap {
  "__metadata": Object {
    "__cache": Object {
      "@cyclus/SystemMap/BUILT_ORDER": Array [
        "scheduler",
        "database",
        "exampleComponent",
      ],
    },
  },
  "map": Object {
    "database": NewDatabase {
      "__metadata": Object {
        "dependencies": Object {},
        "isInitialised": true,
      },
      "dbConnection": "OPENED",
    },
    "exampleComponent": ExampleComponent {
      "__metadata": Object {
        "dependencies": Object {
          "database": "database",
          "scheduler": "scheduler",
        },
        "isInitialised": true,
      },
      "database": NewDatabase {
        "__metadata": Object {
          "dependencies": Object {},
          "isInitialised": true,
        },
        "dbConnection": "OPENED",
      },
      "scheduler": Scheduler {
        "__metadata": Object {
          "dependencies": Object {},
          "isInitialised": true,
        },
        "tick": 10,
      },
    },
    "scheduler": Scheduler {
      "__metadata": Object {
        "dependencies": Object {},
        "isInitialised": true,
      },
      "tick": 10,
    },
  },
}
`;

exports[`cyclus replacing dependencies on the fly 'should work correctly 3`] = `
Array [
  "Start Scheduler",
  "Start Database",
  "Start ExampleComponent",
  "Stop Database",
  "Start NewDatabase",
]
`;

exports[`cyclus simple component should able to replace the config on the fly 1`] = `
SystemMap {
  "__metadata": Object {
    "__cache": Object {
      "@cyclus/SystemMap/BUILT_ORDER": Array [
        "config",
        "component",
      ],
    },
  },
  "map": Object {
    "component": Component {
      "__metadata": Object {
        "dependencies": Object {
          "config": "config",
        },
        "isInitialised": true,
      },
      "config": Object {
        "c": 3,
        "d": 4,
      },
    },
    "config": Object {
      "c": 3,
      "d": 4,
    },
  },
}
`;

exports[`cyclus simple component should work correctly 1`] = `
SystemMap {
  "__metadata": Object {
    "__cache": Object {
      "@cyclus/SystemMap/BUILT_ORDER": Array [
        "config",
        "component",
      ],
    },
  },
  "map": Object {
    "component": Component {
      "__metadata": Object {
        "dependencies": Object {
          "config": "config",
        },
        "isInitialised": true,
      },
      "config": Object {
        "a": 1,
        "b": 2,
      },
    },
    "config": Object {
      "a": 1,
      "b": 2,
    },
  },
}
`;
